cylc/flow/task_pool.py

```diff
@@ -1600,62 +1605,184 @@ def spawn_task(
+        """Set prerequisites or outputs of target tasks.
+
+        Default: set all required outputs.
+
+        Set prerequisites:
+        - spawn the task (if not spawned)
+        - update its prerequisites
+
+        Set outputs:
+        - update task outputs in the DB
+        - spawn children of the outputs (if not spawned)
+        - update the child prerequisites
+        - implied outputs are handled universally by the event manager
+
+        Uses a transient task proxy to spawn children. (Even if parent was
+        previously spawned in this flow its children might not have been).
 
-        User-facing command name: set_outputs. Creates a transient parent just
-        for the purpose of spawning children.
+        Task matching restrictions (for now):
+        - globs (cycle and name) only match in the pool
+        - future tasks must be specified individually
+        - family names are not expanded to members
 
         Args:
-            items: Identifiers for matching task definitions, each with the
-                form "point/name".
-            outputs: List of outputs to spawn on
-            flow_num: Flow number to attribute the outputs
+            items: task ID match patterns
+            prereqs: prerequisites to set
+            outputs: outputs to set and spawn children of
+            flow: flow numbers for spawned or merged tasks
+            flow_wait: wait for flows to catch up before continuing
+            flow_descr: description of new flow
 
         """
-        outputs = outputs or [TASK_OUTPUT_SUCCEEDED]
-        if flow_num is None:
-            flow_nums = None
+        flow_nums = self._get_flow_nums(flow, flow_descr)
+        if flow_nums is None:
+            # Illegal flow command opts
+            return
+
+        # Get matching pool tasks and future task definitions.
+        itasks, future_tasks, unmatched = self.filter_task_proxies(
+            items,
+            future=True,
+            warn=False,
+        )
+
+        for itask in itasks:
+            self.merge_flows(itask, flow_nums)
+            if prereqs:
+                self._set_prereqs_itask(
+                    itask, prereqs, flow_nums, flow_wait)
+            else:
+                self._set_outputs_itask(itask, outputs)
+
+        for name, point in future_tasks:
+            tdef = self.config.get_taskdef(name)
+            if prereqs:
+                self._set_prereqs_tdef(
+                    point, tdef, prereqs, flow_nums, flow_wait)
+            else:
+                trans = self._spawn_transient(
+                    point, tdef, flow_nums, flow_wait)
+                if trans is not None:
+                    self._set_outputs_itask(trans, outputs)
+
+        if self.compute_runahead():
+            self.release_runahead_tasks()
+
+    def _set_outputs_itask(
+        self,
+        itask: 'TaskProxy',
+        outputs: Optional[List[str]],
```

Comment:
[comment] FYI: If you're only iterating over something (not adding, removing or sorting), consider using iterable rather than list.

This makes it flexible to swap to tuples, sets, generators, etc.

```suggestion
        outputs: Optional[Iterable[str]],
```